src/storage/mvcc/consistency_check.rs

```diff
@@ -0,0 +1,462 @@
+    fn update_context(&self, context: &mut Vec<u8>) -> bool {
+        context.push(ConsistencyCheckMethod::Mvcc as u8);
+        context.reserve(8);
+        let len = context.len();
+
+        let mut safe_point = self.local_safe_point.load(AtomicOrdering::Acquire);
+        safe_point = get_safe_point_for_check(safe_point);
+        unsafe {
+            context.set_len(len + 8);
+            std::ptr::copy_nonoverlapping(
+                safe_point.to_le_bytes().as_ptr(),
+                &mut context[len] as _,
+                8,
+            );
+        }
+        // Skiped all other observers.
+        true
+    }
+
+    fn compute_hash(
+        &self,
+        region: &kvproto::metapb::Region,
+        context: &mut &[u8],
+        snap: &E::Snapshot,
+    ) -> Result<Option<u32>> {
+        if context.is_empty() {
+            return Ok(None);
+        }
+        assert_eq!(context[0], ConsistencyCheckMethod::Mvcc as u8);
+        let safe_point = u64::from_le_bytes(context[1..9].try_into().unwrap());
+        *context = &context[9..];
+
+        let local_safe_point = self.local_safe_point.load(AtomicOrdering::Acquire);
+        if safe_point < local_safe_point {
+            warn!(
+                "skip consistency check"; "region_id" => region.get_id(),
+                "safe_ponit" => safe_point,
+                "local_safe_point" => local_safe_point,
+            );
+            return Ok(None);
+        }
+
+        let mut scanner = MvccInfoScanner::new(
+            |cf, opts| snap.iterator_cf_opt(cf, opts).map_err(|e| box_err!(e)),
+            Some(&keys::data_key(region.get_start_key())),
+            Some(&keys::data_end_key(region.get_end_key())),
+            MvccChecksum::new(safe_point),
+        )?;
+        while scanner.next_item()?.is_some() {}
+
+        // Computes the hash from the Region state too.
+        let mut digest = scanner.observer.digest;
+        let region_state_key = keys::region_state_key(region.get_id());
+        digest.update(&region_state_key);
+        match snap.get_value_cf(CF_RAFT, &region_state_key) {
+            Err(e) => return Err(e.into()),
+            Ok(Some(v)) => digest.update(&v),
+            Ok(None) => {}
+        }
+        Ok(Some(digest.finalize()))
+    }
+}
+
+pub trait MvccInfoObserver {
+    type Target;
+
+    // Meet a new mvcc record prefixed `key`.
+    fn on_new_item(&mut self, key: &[u8]);
+    // Emit a complete mvcc record.
+    fn emit(&mut self) -> Self::Target;
+
+    fn on_write(&mut self, key: &[u8], value: &[u8]) -> Result<bool>;
+    fn on_lock(&mut self, key: &[u8], value: &[u8]) -> Result<bool>;
+    fn on_default(&mut self, key: &[u8], value: &[u8]) -> Result<bool>;
+}
+
+pub struct MvccInfoScanner<Iter: EngineIterator, Ob: MvccInfoObserver> {
+    lock_iter: Iter,
+    default_iter: Iter,
+    write_iter: Iter,
+    observer: Ob,
+}
+
+impl<Iter: EngineIterator, Ob: MvccInfoObserver> MvccInfoScanner<Iter, Ob> {
```

Comment:
Could we calculate hash value for `lockcf` and `writecf` independently?

Reply:
It's possible, but filter them by safe point is stil necessary. But for default CF, it needs some information that collected when scanning write CF.